go/libraries/utils/gpg/sign.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gpg
+
+import (
+	"bytes"
+	"context"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"log"
+	"os/exec"
+	"strings"
+
+	"golang.org/x/sync/errgroup"
+)
+
+func execGpgAndReadOutput(ctx context.Context, in []byte, args []string) (*bytes.Buffer, *bytes.Buffer, error) {
+	cmdStr := fmt.Sprintf("gpg %s", strings.Join(args, " "))
+	cmd := exec.CommandContext(ctx, "gpg", args...)
+
+	stdOut, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get stdout for command '%s': %w", cmdStr, err)
+	}
+
+	stdErr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get stderr for command '%s': %w", cmdStr, err)
+	}
+
+	stdIn, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get stdin for command '%s': %w", cmdStr, err)
+	}
+
+	err = cmd.Start()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to start command '%s': %w", cmdStr, err)
+	}
+
+	eg, egCtx := errgroup.WithContext(ctx)
+	outBuf := listenToOut(egCtx, eg, stdOut)
+	errBuf := listenToOut(egCtx, eg, stdErr)
+
+	n, err := io.Copy(stdIn, strings.NewReader(string(in)))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to write stdin for command '%s': %w", cmdStr, err)
+	} else if n < int64(len(in)) {
+		return nil, nil, fmt.Errorf("failed to write stdin for command '%s': EOF before all bytes written", cmd)
+	}
+
+	err = stdIn.Close()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to close stdin for command '%s': %w", cmdStr, err)
+	}
+
+	for {
+		state, err := cmd.Process.Wait()
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to wait for command '%s': %w", cmdStr, err)
+		}
+
+		if state.Exited() {
+			if state.ExitCode() != 0 {
+				waitErr := eg.Wait()
+				if waitErr != nil {
+					return nil, nil, fmt.Errorf("failed to read output for command '%s': %w", cmdStr, waitErr)
+				}
+
+				return nil, nil, fmt.Errorf("command '%s' exited with code %d. stdout: '%s', stderr: '%s'", cmdStr, state.ExitCode(), outBuf.String(), errBuf.String())
+			}
+
+			break
+		}
+	}
+
+	err = eg.Wait()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to read output for command '%s': %w", cmdStr, err)
+	}
+
+	return outBuf, errBuf, nil
+}
+
+func ImportKey(ctx context.Context, keyFile string) error {
```

Comment:
Should have brief docs for all these funcs

Reply:
done